src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const history = [];
+
+  for (const i of actions) {
+    switch (i.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...i.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of i.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+      default:
+        throw new Error(`Unknown action type: ${i.type}`);
+    }
+
+    history.push(currentState);
```

Comment:
[DONT REPEAT YOURSELF] - It's better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed. This way, you ensure that the original `currentState` is not being mutated by reference in the history array.